File: docs/dp_tabulation/1510_hard/perfect_stone_game.cpp
#include <vector>
using namespace std;

bool judgeWinnability(int stonesCount) // LeetCode Q.1510.
{
    vector<int> squaresCounts(stonesCount + 1, 0);

    for (int num = 1; num <= stonesCount; num++) {
        int maxSqrt = static_cast<int>(pow(num, 0.5));

        if (pow(maxSqrt, 2) == num) {
            squaresCounts[num] += 1; // From 0 to 1.
            continue;
        }

        for (int sqrt = maxSqrt; sqrt >= 1; sqrt--) {
            int residual = num - pow(sqrt, 2);
            if (squaresCounts[residual] == 0) {
                squaresCounts[num] += 1; // From 0 to 1.
                break;
            }
        }
    }

    return squaresCounts.back() == 1;
}

File: docs/dp_tabulation/1510_hard/perfect_stone_game.py

def judge_winnability(stones_count: int) -> bool:  # LeetCode Q.1510.
    squares_counts = [0] * (stones_count + 1)

    for num in range(1, stones_count + 1):
        max_sqrt = int(num ** 0.5)
        if max_sqrt ** 2 == num:
            squares_counts[num] += 1  # From 0 to 1.
            continue

        for sqrt in range(max_sqrt, 0, -1):
            residual = num - (sqrt ** 2)
            if squares_counts[residual] == 0:
                squares_counts[num] += 1  # From 0 to 1.
                break

    return squares_counts[-1] == 1
